### PR TITLE
automatically download and upload circleci wheels

### PR DESCRIPTION
**Pull request by reaperhulk**
partially tested but not end to end. I need to dig up an offset where I can grab a pipeline with a tag to see if it all works.

### CODE REVIEW

File: release.py [hunk 1]
```diff
@@ -107,6 +203,7 @@ def release(version):
         f"description={version}&scopes=repo"
     )
     github_token = getpass.getpass("Github person access token: ")
+    circle_token = getpass.getpass("CircleCI token: ")
```
Comment (alex) [inline release.py:206]: Can we print instructions on where to get one :-)

File: release.py [hunk 2]
```diff
@@ -95,6 +95,102 @@ def fetch_github_actions_artifacts(token, version):
...
+    pipeline_id = None
+    for item in response.json()["items"]:
+        if item["project_slug"] == "gh/pyca/cryptography":
+            if item["vcs"].get("tag", None) == version:
+                pipeline_id = item["id"]
+                break
+
+    if pipeline_id is None:
+        raise ValueError(
+            "Could not find a pipeline for version {0}".format(version)
```
Comment (alex) [inline release.py:148]:
```suggestion
            f"Could not find a pipeline for version {version}"
```

File: release.py [hunk 3]
```diff
@@ -95,6 +95,102 @@ def fetch_github_actions_artifacts(token, version):
...
+def wait_for_build_complete_circleci(session, token, pipeline_id):
+    while True:
+        response = session.get(
+            f"https://circleci.com/api/v2/pipeline/{pipeline_id}/workflow",
+            headers={
+                "Circle-Token": token,
+            },
+        )
+        response.raise_for_status()
+        if response.json()["status"] == "success":
```
Comment (alex) [inline release.py:107]: This should probably do something besides spin forever if it fails, yeah?

**Comment by reaperhulk** [comment]
(Comments addressed, everything tested except the == version match on the tag now)

File: release.py [hunk 4]
```diff
@@ -95,6 +95,103 @@ def fetch_github_actions_artifacts(token, version):
...
+            f"https://circleci.com/api/v2/pipeline/{pipeline_id}/workflow",
+            headers={
+                "Circle-Token": token,
+            },
+        )
+        response.raise_for_status()
+        status = response.json()["items"][0]["status"]
+        if status == "success":
+            break
+        elif status != "running":
```
Comment (alex) [inline release.py:110]: Is there a queued status or something like that, before it starts?

File: release.py [hunk 5]
```diff
@@ -95,6 +95,103 @@ def fetch_github_actions_artifacts(token, version):
...
+        if item["project_slug"] == "gh/pyca/cryptography":
+            if item["vcs"].get("tag", None) == version:
+                pipeline_id = item["id"]
+                break
+
+    if pipeline_id is None:
+        raise ValueError(f"Could not find a pipeline for version {version}")
+
+    wait_for_build_complete_circleci(session, token, pipeline_id)
+    urls = fetch_circleci_artifact_urls(session, token, item["id"])
```
Comment (alex) [inline release.py:153]:
```suggestion
    urls = fetch_circleci_artifact_urls(session, token, pipline_id)
```

File: release.py [hunk 6]
```diff
@@ -95,6 +95,103 @@ def fetch_github_actions_artifacts(token, version):
...
+            break
+        elif status != "running":
+            raise ValueError(f"CircleCI build failed with status {status}")
+        time.sleep(3)
+
+
+def download_artifacts_circleci(session, urls):
+    paths = []
+    for url in urls:
+        name = url.split("/")[-1]
```
Comment (alex) [inline release.py:118]:
```suggestion
        name = os.path.basename(urllib.parse.urlparse(url).path)
```

File: release.py [hunk 7]
```diff
@@ -95,6 +95,103 @@ def fetch_github_actions_artifacts(token, version):
...
+        response = session.get(
+            url,
+            headers={},
```
Comment (alex) [inline release.py:121]:
```suggestion
```

File: release.py [hunk 8]
```diff
@@ -95,6 +95,103 @@ def fetch_github_actions_artifacts(token, version):
...
+            raise ValueError(f"CircleCI build failed with status {status}")
+        time.sleep(3)
+
+
+def download_artifacts_circleci(session, urls):
+    paths = []
+    for url in urls:
+        name = url.split("/")[-1]
+        response = session.get(
+            url,
```
Comment (alex) [inline release.py:120]: SSRF? lolsob

**Review by alex — APPROVED** [review]

**Pull request merged**